### PR TITLE
feat: Implement Spark-compatible CAST between integer types

### LINKED ISSUE #311: Implement Spark-compatible CAST between integer types
We currently delegate to DataFusion for casts between integer types, which produces different results when compared to Spark.

For example, when casting from long to short in `LEGACY` mode, Spark will cast `-9197595427105923027` to `8237` and DataFusion will cast to `null` because we are using DataFusion's CastOptions with `safe` mode disabled, which is equivalent to Spark's `TRY` mode.

The behavior in `TRY` mode is correct.

The behavior in `ANSI` mode is basically correct but the error message returned by DataFusion does not match Spark and it isn't practical to try and parse the error string and translate to Spark's error message.

It is probably best for now just to write our own Spark-compatible cast. The following code sample shows the two methods for casting that we would need to use (`try_from` for `TRY` and `ANSI` mode, and `as` for `LEGACY` mode).

```rust
fn main() {
    let v = -9197595427105923027_i64;
    println!("{:?}", i16::try_from(v));
    println!("{:}", v as i16);
}
```

This produces the following output:

```
Err(TryFromIntError(()))
8237
```

### PR DESCRIPTION
**Pull request by ganeshkumar269**
## Which issue does this PR close?

Closes #311 

## What changes are included in this PR?

- handled cases where casting b/w ints causes difference in behaviour b/w datafusion and spark. for eg. long to int, int to byte etc. 
- ignored the cases where the behaviour is normal. for eg. int to long, short to int, all cases of TRY mode.

## How are these changes tested?

added corresponding scala test cases.

### CODE REVIEW

**Comment by ganeshkumar269** [comment]
Hi @viirya @andygrove , firstly please let me know if this PR aligns with the expectations on how to fix the issue, if not kindly provide pointers on how I can move in the right direction.

Also reg the errors in the CI pipeline, looks like error message for 3.2 is a bit different compared to 3.3 and 3.4 for overflow cases. so in the rust code I will have to add a check for spark version and return the error message accordingly. 
How can I get spark version from the rust side?

**Comment by viirya** [comment]
> Also reg the errors in the CI pipeline, looks like error message for 3.2 is a bit different compared to 3.3 and 3.4 for overflow cases. so in the rust code I will have to add a check for spark version and return the error message accordingly.
> How can I get spark version from the rust side?

For this kind of case, we tend to return same error message at native side, but handle this difference in Scala tests.

**Comment by andygrove** [comment]
As @viirya said, we can handle the difference in error message format in the scala test (we already have examples of this in the `castTest` method).

For versions prior to 3.4 perhaps you could just check for the word "overflow" in the error message. 

```
- cast short to byte *** FAILED *** (439 milliseconds)
  "Execution error: [CAST_OVERFLOW] The value 18716S of the type "SMALLINT" cannot be cast to "TINYINT" due to an overflow. Use `try_cast` to tolerate overflow and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error." did not contain "asting 18716 to tinyint causes overflow" (CometCastSuite.scala:200)
```

**Comment by ganeshkumar269** [comment]
thanks for the inputs @viirya @andygrove , I have added another check in the assert statement where we compare exception messages. 
Do you think this approach looks good? or do I need to separate the "invalid cast" and "overflow" assert statements. 

If this change is fine, I will modify the code comments to accommodate "overflow" exception aswell.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 1]
```diff
@@ -722,7 +746,8 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
               // Comet message is in format `The value 'VALUE' of the type FROM_TYPE cannot be cast to TO_TYPE`
               // We just check that the comet message contains the same invalid value as the Spark message
               val sparkInvalidValue = sparkMessage.substring(sparkMessage.indexOf(':') + 2)
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:748]: Now that we are handling multiple types of error here, we should probably check if `sparkMessage.indexOf(':')` returns a non-zero value before trying to use it.

**Comment by andygrove** [comment]
> thanks for the inputs @viirya @andygrove , I have added another check in the assert statement where we compare exception messages. Do you think this approach looks good? or do I need to separate the "invalid cast" and "overflow" assert statements.
> 
> If this change is fine, I will modify the code comments to accommodate "overflow" exception aswell.

I think this general approach is OK to handle different types of expected errors. I left a specific comment on the code as well.

**Comment by ganeshkumar269** [comment]
Hi @andygrove , i have added a check before we fetch sparkInvalidValue, defaulting it to EMPTY_STRING if ':' is not present.
Also added additional comments on why we are checking for the presence of 'overflow' string.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 2]
```diff
@@ -807,11 +833,22 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
+              // Comet message is in format `The value 'VALUE' of the type FROM_TYPE cannot be cast to TO_TYPE
+              // due to an overflow`
+              // We check if the comet message contains 'overflow'.
+              val sparkInvalidValue = if (sparkMessage.indexOf(':') == -1) {
+                EMPTY_STRING
+              } else {
+                sparkMessage.substring(sparkMessage.indexOf(':') + 2)
+              }
+              assert(
+                cometMessage.contains(sparkInvalidValue) || cometMessage.contains("overflow"))
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:851]: If `sparkInvalidValue` is `EMPTY_STRING`, won't `cometMessage.contains(sparkInvalidValue)` always be true?

Comment (ganeshkumar269) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:851]: you are right, my bad 😅 . so incase sparkMessage doesnt have ':' should I assert on just commetMessage.contains("overflow")
something like this,

```
if sparkMessage.indexOf(':') == -1 then assert(commetMessage.contains("overflow"))
else assert(commetMessage.contains(sparkInvalidValue) || commetMessage.contains("overflow"))
```

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:851]: Yes, something like that. I haven't reviewed the overflow messages to see if they contain `:` though (in any of the spark versions 3.2, 3.3, and 3.4)

Comment (ganeshkumar269) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:851]: doesnt look like overflow error message has ':' in it.

**3.4** - [CAST_OVERFLOW] The value 9223372036854775807L of the type "BIGINT" cannot be cast to "INT" due to an overflow. Use `try_cast` to tolerate overflow and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error
**3.3** - The value 9223372036854775807L of the type "BIGINT" cannot be cast to "INT" due to an overflow. Use `try_cast` to tolerate overflow and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error
**3.2** - Casting 9223372036854775807 to int causes overflow

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 3]
```diff
@@ -665,6 +665,30 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     castTest(generateTimestamps(), DataTypes.DateType)
   }
 
+  test("cast short to byte") {
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:668]: These methods already exist in main but have different naming, so I think you need to upmerge/rebase against main.

Example:

```
ignore("cast ShortType to ByteType")
```

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @ganeshkumar269

**Comment by ganeshkumar269** [comment]
thanks @andygrove 🙏🏾  , glad to be a contributor to comet.

**Comment by andygrove** [comment]
@ganeshkumar269 It looks like the error message check needs a little more work. Some tests are failing on Spark 3.3.

This error message does not contain `:` or `overflow`. In this specific case, looking for `CAST_INVALID_INPUT` would be a more robust check.

```
- cast StringType to LongType *** FAILED *** (416 milliseconds)
  "[CAST_INVALID_INPUT] The value '-9223372036854775809' of the type "STRING" cannot be cast to "BIGINT" because it is malformed. Correct the value as per the syntax, or change its target type. Use `try_cast` to tolerate malformed input and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error." 
did not contain "overflow" (CometCastSuite.scala:881)
  ```

**Comment by andygrove** [comment]
@ganeshkumar269 My original code for comparing errors in 3.2/3.3 was not very robust. I am also looking at this now to see if I can help improve these checks.

**Comment by ganeshkumar269** [comment]
> @ganeshkumar269 My original code for comparing errors in 3.2/3.3 was not very robust. I am also looking at this now to see if I can help improve these checks.

is there a way I could help here?

**Comment by andygrove** [comment]
@ganeshkumar269 here is my suggestion:

```scala
            if (CometSparkSessionExtensions.isSpark34Plus) {
              // for Spark 3.4 we expect to reproduce the error message exactly
              assert(cometMessage == sparkMessage)
            } else if (CometSparkSessionExtensions.isSpark33Plus) {
              // for Spark 3.3 we just need to strip the prefix from the Comet message
              // before comparing
              val cometMessageModified = cometMessage
                .replace("[CAST_INVALID_INPUT] ", "")
                .replace("[CAST_OVERFLOW] ", "")
              assert(cometMessageModified == sparkMessage)
            } else if (CometSparkSessionExtensions.isSpark32) {
              // for Spark 3.2 we just make sure we are seeing a similar type of error
              if (sparkMessage.contains("causes overflow")) {
                assert(cometMessage.contains("due to an overflow"))
              } else {
                // assume that this is an invalid input message in the form:
                // `invalid input syntax for type numeric: -9223372036854775809`
                // we just check that the Comet message contains the same literal value
                val i = sparkMessage.indexOf(':') + 2
                assert(cometMessage.contains(sparkMessage.substring(i)))
              }
            }
```

I tested this from the command line for all Spark versions using:

```
mvn test -Pspark-3.2 -DwildcardSuites=org.apache.comet.CometCastSuite -Dspotless.check.skip=true
```

**Review by andygrove — APPROVED** [review]

**Pull request merged**